### PR TITLE
Add kml:Link, kml:Icon

### LINKED ISSUE #10: Add kml:Link, kml:Icon
These two can be done together (but should be separate structs) and are [13.1 in the specification](http://docs.opengeospatial.org/is/12-007r2/12-007r2.html#974). 

This can be based on the implementation for `kml:Location` in #7

### PR DESCRIPTION
**Pull request by k-mack**
Attempt to add support for `kml:Link` and `kml:Icon`.

A couple of points to discuss.

### First

Since the spec's `kml:Icon` (or `kml:BasicLinkType`) is already implemented and exported by the library as `Kml::Icon`, I used an enum to model the spec's `kml:LinkType`. Its variants are `Kml::Link` and `Kml:Icon`. Both are constructed with `kml::types::LinkModel`, the common data model used by the spec's `kml:LinkType`.

If `kml:Model` was to be implemented using this PR, then it's `link` field would be of type `kml::types::LinkType` and it would need to be guarded so as not to have it be set to `kml::types::LinkType::Icon`:

```rust
pub struct Model {
    link: Option<kml::types::LinkType>,
    // ...
}

impl Default for Model {
    fn default() -> Model {
        Model { link: None }
    }
}

impl Model {
    pub fn get_link(&self) -> Option<&kml::types::LinkType> {
        self.link.as_ref()
    }

    pub fn set_link(&mut self, link_model: &kml::types::LinkModel) {
        self.link = Some(kml::types::LinkType::Link(link_model.clone()));
    }
}
```

Let me know how you feel about this approach.

Alternatively, I could use two identical structs with different names. I could expose Icon as `Kml::LinkIcon` to not have it interfere with the existing `Kml::Icon`.

What makes the most sense for an approachable, intuitive API?

### Second

I used empty strings for the optional String fields that have no default value. Let me know if you'd prefer to have them be `Option`s instead.

Closes #10.

### CODE REVIEW

File: src/writer.rs [hunk 1]
```diff
@@ -427,6 +428,33 @@ where
             .write_event(Event::End(BytesEnd::borrowed(b"ListStyle")))?)
     }
 
+    fn write_link_type(&mut self, link: &LinkType) -> Result<(), Error> {
+        let (tag_name, link_model) = match link {
+            LinkType::Icon(model) => (b"Icon", &*model),
+            LinkType::Link(model) => (b"Link", &*model),
+        };
+        self.writer
+            .write_event(Event::Start(BytesStart::owned_name(tag_name.to_vec())))?;
```
Comment (pjsier) [inline src/writer.rs:437]: We'll need to add this here to capture the attributes

```rust
.with_attributes(self.hash_map_as_attrs(&link_model.attrs))
```

**Review by pjsier — COMMENTED** [review]
Thanks so much for the PR and thorough write up! I think this approach makes sense, but I'm a bit worried about needing a setter for `kml:Model`. Other than the duplicate code do you see any issues creating separate structs?

Otherwise it would be great to convert the fields with empty string to `Option<String>` since that's what we're using for other elements, and thanks for flagging that.

**Comment by k-mack** [comment]
As much as I like the idea of using an enum, it doesn't make sense from an API ergonomics perspective. If Rust supported enum variants as first-class types, then I think it'd be a clean approach.

I'll break the enum into separate structs.

File: src/writer.rs (hunk 1 above)
Comment (k-mack) [reply src/writer.rs]: Done.

File: src/reader.rs [hunk 2]
```diff
@@ -930,6 +1031,64 @@ mod tests {
...
+
+        let l: Kml = kml_str.parse().unwrap();
+        assert_eq!(
+            l,
+            Kml::LinkTypeLink(LinkTypeLink {
+                href: Some("/path/to/local/resource".to_string()),
+                refresh_mode: Some(types::RefreshMode::OnChange),
+                view_refresh_mode: Some(types::ViewRefreshMode::OnStop),
+                view_format: Some(String::new()),
+                attrs,
```
Comment (k-mack) [inline src/reader.rs:1057]: Do we want to store the attributes? I see some other types do, but most ignore them. The spec only allows three here: `id`, `targetId`, and a user-defined one. Maybe I should ignore them to conform to how it is done for other types?

As an aside, I could see how _many_ attributes could be an issue with memory. Perhaps that's a separate issue to be solved in another PR.

File: src/reader.rs [hunk 3]
```diff
@@ -930,6 +1031,64 @@ mod tests {
...
+        attrs.insert("id".to_string(), "Some ID".to_string());
+
+        let l: Kml = kml_str.parse().unwrap();
+        assert_eq!(
+            l,
+            Kml::LinkTypeLink(LinkTypeLink {
+                href: Some("/path/to/local/resource".to_string()),
+                refresh_mode: Some(types::RefreshMode::OnChange),
+                view_refresh_mode: Some(types::ViewRefreshMode::OnStop),
+                view_format: Some(String::new()),
```
Comment (k-mack) [inline src/reader.rs:1056]: In this test, `viewFormat` is an empty element, and this is treated as a value. Any desire to identify this and assign the optional String value to `None`?

File: src/types/mod.rs [hunk 4]
```diff
@@ -36,6 +36,10 @@ mod geometry;
 
 pub use geometry::Geometry;
 
+mod link;
+
+pub use link::{Icon as LinkTypeIcon, Link as LinkTypeLink, RefreshMode, ViewRefreshMode};
```
Comment (k-mack) [inline src/types/mod.rs:41]: Not a fan of these aliases, but I couldn't think of anything better. Decided to alias `Link` only because `Icon` was aliased. WDYT?

**Review by k-mack — COMMENTED** [review]
Separate structs feels more Rusty. I left a couple of comments to call out some of the decisions I made that are worth a review.

File: src/reader.rs (hunk 2 above)
Comment (pjsier) [reply src/reader.rs:1057]: I think enough of the other types include `attrs` to keep it for now, but worth flagging for the future

File: src/reader.rs (hunk 3 above)
Comment (pjsier) [reply src/reader.rs:1056]: I think the test and behavior work as-is, since I would assume `None` would mean the tag isn't present at all

File: src/types/mod.rs (hunk 4 above)
Comment (pjsier) [reply src/types/mod.rs:41]: I think I'd prefer we only alias `Icon` to `LinkTypeIcon` and remove the `LinkTypeLink` to keep things as consistent with the spec as possible

File: src/reader.rs [hunk 5]
```diff
@@ -599,6 +604,102 @@ where
...
+                Event::Start(ref mut e) => match e.local_name() {
+                    b"href" => icon.href = Some(self.read_str()?),
+                    b"refreshMode" => {
+                        icon.refresh_mode = match self.read_str()?.as_str() {
+                            "onChange" => Some(RefreshMode::OnChange),
+                            "onInterval" => Some(RefreshMode::OnInterval),
+                            "onExpire" => Some(RefreshMode::OnExpire),
+                            _ => None,
+                        }
+                    }
```
Comment (pjsier) [inline src/reader.rs:627]: Instead of including this here, could we implement `FromStr` for `RefreshMode` as well as the other added enums? We've done this for `AltitudeMode` and `ColorMode` as well and it keeps things a little cleaner

File: src/types/link.rs [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+            RefreshMode::OnChange => write!(f, "onChange"),
+            RefreshMode::OnInterval => write!(f, "onInterval"),
+            RefreshMode::OnExpire => write!(f, "onExpire"),
+        }
+    }
+}
+
+/// `kml:viewRefreshModeEnumType`, [16.27](https://docs.opengeospatial.org/is/12-007r2/12-007r2.html#1270) in the KML specification.
+#[derive(Clone, Debug, PartialEq)]
+pub enum ViewRefreshMode {
```
Comment (pjsier) [inline src/types/link.rs:92]: It would be great to implement `FromStr` here as well

**Review by pjsier — COMMENTED** [review]
This is looking great, thanks for the changes! The last two things to update would be implementing `FromStr` for the enums and removing the `Link` alias, then I think this is ready to merge

File: src/reader.rs (hunk 5 above)
Comment (k-mack) [reply src/reader.rs]: Done.

File: src/types/link.rs (hunk 6 above)
Comment (k-mack) [reply src/types/link.rs:106]: Done.

**Review by pjsier — APPROVED** [review]
This looks great, thanks again!
